### ISSUE TITLE
Setup tun failed to mount permission

### ISSUE THREAD

**Issue by cryptofuture** [lxc/lxd#2185]
I tried setup tun for lxd container using [this](https://askubuntu.com/questions/650471/how-do-i-allow-dev-net-tun-in-a-lxd-managed-lxc-container) trick:

```
lxc config device add mycontainername tun unix-char path=/dev/net/tun
```
And it works (with running container), but I can't start/restart container with tun enabled.
```
  lxc 20160707143151.710 ERROR    lxc_utils - utils.c:safe_mount:1692 - Permission denied - Failed to mount /var/lib/lxd/devices/mycontainername/unix.dev-net-tun onto /usr/lib/x86_64-linux-gnu/lxc/dev/net/tun
            lxc 20160707143151.710 ERROR    lxc_conf - conf.c:mount_entry:1650 - Permission denied - failed to mount '/var/lib/lxd/devices/mycontainername/unix.dev-net-tun' on '/usr/lib/x86_64-linux-gnu/lxc/dev/net/tun'
            lxc 20160707143151.710 ERROR    lxc_conf - conf.c:lxc_setup:3731 - failed to setup the mount entries for 'mycontainername'
            lxc 20160707143151.710 ERROR    lxc_start - start.c:do_start:833 - failed to setup the container
            lxc 20160707143151.710 ERROR    lxc_sync - sync.c:__sync_wait:57 - An error occurred in another process (expected sequence number 3)
            lxc 20160707143151.710 WARN     lxc_conf - conf.c:lxc_delete_network:2909 - failed to remove interface 18 'eth0'
            lxc 20160707143151.710 ERROR    lxc_start - start.c:__lxc_start:1353 - failed to spawn 'mycontainername'

```

### Required information

 * Distribution: Xenial
 * Distribution version: 16.04
 * The output of "lxc info":
   
```
 driver: lxc
  driverversion: 2.0.3
  kernel: Linux
  kernelarchitecture: x86_64
  kernelversion: 4.4.0-30-generic
  server: lxd
  serverpid: 10305
  serverversion: 2.0.3
  storage: dir
  storageversion: ""
config: {}
public: false
```
Any right way to setup tun for lxd container?

**Comment by cryptofuture** [comment]
Same error with version from ppa:ubuntu-lxc/lxd-git-master. Just tested.

**Comment by stgraber** [comment]
```
stgraber@castiana:~$ lxc start mycontainer
stgraber@castiana:~$ lxc config device add mycontainer tun unix-char path=/dev/net/tun
Device tun added to mycontainer
stgraber@castiana:~$ lxc exec mycontainer -- ls -lh /dev/net/tun
crw-rw---- 1 root root 10, 200 Jul  7 17:36 /dev/net/tun
stgraber@castiana:~$ 
```

Also, current git master actually gives you /dev/net/tun by default.

**Comment by cryptofuture** [comment]
It works on running container. And 
```
lxc exec mycontainer -- ls -lh /dev/net/tun
```
Shows tun exist, but after container restart, I can't just start it at all, same log as at the begining of the post.
```
error: Error calling 'lxd forkstart con1 /var/lib/lxd/containers /var/log/lxd/mycontainername/lxc.conf': err='exit status 1'
Try `lxc info --show-log mycontainername` for more info
```
btw /var/lib/lxd/devices/mycontainername/ path is empty.

```
Name: mycontainername
Architecture: i686
Status: Stopped
Type: persistent
Profiles: default

Log:

            lxc 20160707201003.622 INFO     lxc_start - start.c:lxc_check_inherited:251 - closed inherited fd 3
            lxc 20160707201003.622 INFO     lxc_start - start.c:lxc_check_inherited:251 - closed inherited fd 9
            lxc 20160707201003.625 INFO     lxc_container - lxccontainer.c:do_lxcapi_start:797 - Attempting to set proc title to [lxc monitor] /var/lib/lxd/containers mycontainername
            lxc 20160707201003.626 INFO     lxc_start - start.c:lxc_check_inherited:251 - closed inherited fd 9
            lxc 20160707201003.626 INFO     lxc_lsm - lsm/lsm.c:lsm_init:48 - LSM security driver AppArmor
            lxc 20160707201003.626 INFO     lxc_seccomp - seccomp.c:parse_config_v2:342 - processing: .reject_force_umount  # comment this to allow umount -f;  not recommended.
            lxc 20160707201003.626 INFO     lxc_seccomp - seccomp.c:parse_config_v2:446 - Adding native rule for reject_force_umount action 0
            lxc 20160707201003.626 INFO     lxc_seccomp - seccomp.c:do_resolve_add_rule:216 - Setting seccomp rule to reject force umounts

            lxc 20160707201003.626 INFO     lxc_seccomp - seccomp.c:parse_config_v2:449 - Adding compat rule for reject_force_umount action 0
            lxc 20160707201003.626 INFO     lxc_seccomp - seccomp.c:do_resolve_add_rule:216 - Setting seccomp rule to reject force umounts

            lxc 20160707201003.626 INFO     lxc_seccomp - seccomp.c:parse_config_v2:342 - processing: .[all].
            lxc 20160707201003.626 INFO     lxc_seccomp - seccomp.c:parse_config_v2:342 - processing: .kexec_load errno 1.
            lxc 20160707201003.626 INFO     lxc_seccomp - seccomp.c:parse_config_v2:446 - Adding native rule for kexec_load action 327681
            lxc 20160707201003.626 INFO     lxc_seccomp - seccomp.c:parse_config_v2:449 - Adding compat rule for kexec_load action 327681
            lxc 20160707201003.626 INFO     lxc_seccomp - seccomp.c:parse_config_v2:342 - processing: .open_by_handle_at errno 1.
            lxc 20160707201003.626 INFO     lxc_seccomp - seccomp.c:parse_config_v2:446 - Adding native rule for open_by_handle_at action 327681
            lxc 20160707201003.626 INFO     lxc_seccomp - seccomp.c:parse_config_v2:449 - Adding compat rule for open_by_handle_at action 327681
            lxc 20160707201003.626 INFO     lxc_seccomp - seccomp.c:parse_config_v2:342 - processing: .init_module errno 1.
            lxc 20160707201003.626 INFO     lxc_seccomp - seccomp.c:parse_config_v2:446 - Adding native rule for init_module action 327681
            lxc 20160707201003.626 INFO     lxc_seccomp - seccomp.c:parse_config_v2:449 - Adding compat rule for init_module action 327681
            lxc 20160707201003.626 INFO     lxc_seccomp - seccomp.c:parse_config_v2:342 - processing: .finit_module errno 1.
            lxc 20160707201003.626 INFO     lxc_seccomp - seccomp.c:parse_config_v2:446 - Adding native rule for finit_module action 327681
            lxc 20160707201003.626 INFO     lxc_seccomp - seccomp.c:parse_config_v2:449 - Adding compat rule for finit_module action 327681
            lxc 20160707201003.626 INFO     lxc_seccomp - seccomp.c:parse_config_v2:342 - processing: .delete_module errno 1.
            lxc 20160707201003.626 INFO     lxc_seccomp - seccomp.c:parse_config_v2:446 - Adding native rule for delete_module action 327681
            lxc 20160707201003.626 INFO     lxc_seccomp - seccomp.c:parse_config_v2:449 - Adding compat rule for delete_module action 327681
            lxc 20160707201003.626 INFO     lxc_seccomp - seccomp.c:parse_config_v2:456 - Merging in the compat seccomp ctx into the main one
            lxc 20160707201003.627 INFO     lxc_conf - conf.c:run_script_argv:367 - Executing script '/usr/bin/lxd callhook /var/lib/lxd 2 start' for container 'mycontainername', config section 'lxc'
            lxc 20160707201003.627 INFO     lxc_start - start.c:lxc_check_inherited:251 - closed inherited fd 3
            lxc 20160707201003.627 INFO     lxc_start - start.c:lxc_check_inherited:251 - closed inherited fd 9
            lxc 20160707201003.630 INFO     lxc_monitor - monitor.c:lxc_monitor_sock_name:178 - using monitor sock name lxc/d78a9d7e97b4b375//var/lib/lxd/containers
            lxc 20160707201003.649 DEBUG    lxc_start - start.c:setup_signal_fd:289 - sigchild handler set
            lxc 20160707201003.649 DEBUG    lxc_console - console.c:lxc_console_peer_default:469 - no console peer
            lxc 20160707201003.649 INFO     lxc_start - start.c:lxc_init:488 - 'mycontainername' is initialized
            lxc 20160707201003.650 INFO     lxc_confile - confile.c:config_idmap:1500 - read uid map: type u nsid 0 hostid 100000 range 65536
            lxc 20160707201003.650 INFO     lxc_confile - confile.c:config_idmap:1500 - read uid map: type g nsid 0 hostid 100000 range 65536
            lxc 20160707201003.650 DEBUG    lxc_start - start.c:__lxc_start:1326 - Not dropping cap_sys_boot or watching utmp
            lxc 20160707201003.650 INFO     lxc_start - start.c:resolve_clone_flags:1013 - Cloning a new user namespace
            lxc 20160707201003.654 DEBUG    lxc_conf - conf.c:instantiate_veth:2614 - instantiated veth 'vethOBYIRR/veth1SNJMI', index is '18'
            lxc 20160707201003.654 INFO     lxc_cgroup - cgroup.c:cgroup_init:68 - cgroup driver cgroupfs-ng initing for mycontainername
            lxc 20160707201003.671 DEBUG    lxc_cgfsng - cgfsng.c:cgfsng_setup_limits:1667 - cgroup 'devices.allow' set to 'c 10:200 rwm'
            lxc 20160707201003.671 INFO     lxc_cgfsng - cgfsng.c:cgfsng_setup_limits:1671 - cgroup has been setup
            lxc 20160707201003.716 DEBUG    lxc_conf - conf.c:lxc_assign_network:3048 - move 'eth0' to '5557'
            lxc 20160707201003.716 NOTICE   lxc_start - start.c:do_start:777 - switching to gid/uid 0/0 in new user namespace
            lxc 20160707201003.716 DEBUG    lxc_conf - conf.c:setup_rootfs:1215 - mounted '/var/lib/lxd/containers/mycontainername/rootfs' on '/usr/lib/x86_64-linux-gnu/lxc'
            lxc 20160707201003.716 INFO     lxc_conf - conf.c:setup_utsname:843 - 'mycontainername' hostname has been setup
            lxc 20160707201003.739 DEBUG    lxc_conf - conf.c:setup_hw_addr:2145 - mac address '00:16:3e:0d:e9:76' on 'eth0' has been setup
            lxc 20160707201003.740 DEBUG    lxc_conf - conf.c:setup_netdev:2372 - 'eth0' has been setup
            lxc 20160707201003.740 INFO     lxc_conf - conf.c:setup_network:2393 - network has been setup
            lxc 20160707201003.740 INFO     lxc_conf - conf.c:mount_autodev:1072 - Mounting container /dev
            lxc 20160707201003.740 INFO     lxc_conf - conf.c:mount_autodev:1095 - Mounted tmpfs onto /usr/lib/x86_64-linux-gnu/lxc/dev
            lxc 20160707201003.740 INFO     lxc_conf - conf.c:mount_autodev:1113 - Mounted container /dev
            lxc 20160707201003.740 DEBUG    lxc_conf - conf.c:mount_entry:1657 - remounting /dev/fuse on /usr/lib/x86_64-linux-gnu/lxc/dev/fuse to respect bind or remount options
            lxc 20160707201003.740 DEBUG    lxc_conf - conf.c:mount_entry:1672 - (at remount) flags for /dev/fuse was 4098, required extra flags are 2
            lxc 20160707201003.740 DEBUG    lxc_conf - conf.c:mount_entry:1707 - mounted '/dev/fuse' on '/usr/lib/x86_64-linux-gnu/lxc/dev/fuse', type 'none'
            lxc 20160707201003.740 DEBUG    lxc_conf - conf.c:mount_entry:1657 - remounting /proc/sys/fs/binfmt_misc on /usr/lib/x86_64-linux-gnu/lxc/proc/sys/fs/binfmt_misc to respect bind or remount options
            lxc 20160707201003.740 DEBUG    lxc_conf - conf.c:mount_entry:1672 - (at remount) flags for /proc/sys/fs/binfmt_misc was 4096, required extra flags are 0
            lxc 20160707201003.740 DEBUG    lxc_conf - conf.c:mount_entry:1681 - mountflags already was 20480, skipping remount
            lxc 20160707201003.740 DEBUG    lxc_conf - conf.c:mount_entry:1707 - mounted '/proc/sys/fs/binfmt_misc' on '/usr/lib/x86_64-linux-gnu/lxc/proc/sys/fs/binfmt_misc', type 'none'
            lxc 20160707201003.740 DEBUG    lxc_conf - conf.c:mount_entry:1657 - remounting /sys/firmware/efi/efivars on /usr/lib/x86_64-linux-gnu/lxc/sys/firmware/efi/efivars to respect bind or remount options
            lxc 20160707201003.740 DEBUG    lxc_conf - conf.c:mount_entry:1672 - (at remount) flags for /sys/firmware/efi/efivars was 4110, required extra flags are 14
            lxc 20160707201003.740 DEBUG    lxc_conf - conf.c:mount_entry:1707 - mounted '/sys/firmware/efi/efivars' on '/usr/lib/x86_64-linux-gnu/lxc/sys/firmware/efi/efivars', type 'none'
            lxc 20160707201003.740 DEBUG    lxc_conf - conf.c:mount_entry:1657 - remounting /sys/fs/fuse/connections on /usr/lib/x86_64-linux-gnu/lxc/sys/fs/fuse/connections to respect bind or remount options
            lxc 20160707201003.740 DEBUG    lxc_conf - conf.c:mount_entry:1672 - (at remount) flags for /sys/fs/fuse/connections was 4096, required extra flags are 0
            lxc 20160707201003.740 DEBUG    lxc_conf - conf.c:mount_entry:1681 - mountflags already was 20480, skipping remount
            lxc 20160707201003.740 DEBUG    lxc_conf - conf.c:mount_entry:1707 - mounted '/sys/fs/fuse/connections' on '/usr/lib/x86_64-linux-gnu/lxc/sys/fs/fuse/connections', type 'none'
            lxc 20160707201003.741 DEBUG    lxc_conf - conf.c:mount_entry:1657 - remounting /sys/fs/pstore on /usr/lib/x86_64-linux-gnu/lxc/sys/fs/pstore to respect bind or remount options
            lxc 20160707201003.741 DEBUG    lxc_conf - conf.c:mount_entry:1672 - (at remount) flags for /sys/fs/pstore was 4110, required extra flags are 14
            lxc 20160707201003.741 DEBUG    lxc_conf - conf.c:mount_entry:1707 - mounted '/sys/fs/pstore' on '/usr/lib/x86_64-linux-gnu/lxc/sys/fs/pstore', type 'none'
            lxc 20160707201003.741 DEBUG    lxc_conf - conf.c:mount_entry:1657 - remounting /sys/kernel/debug on /usr/lib/x86_64-linux-gnu/lxc/sys/kernel/debug to respect bind or remount options
            lxc 20160707201003.741 DEBUG    lxc_conf - conf.c:mount_entry:1672 - (at remount) flags for /sys/kernel/debug was 4096, required extra flags are 0
            lxc 20160707201003.741 DEBUG    lxc_conf - conf.c:mount_entry:1681 - mountflags already was 20480, skipping remount
            lxc 20160707201003.741 DEBUG    lxc_conf - conf.c:mount_entry:1707 - mounted '/sys/kernel/debug' on '/usr/lib/x86_64-linux-gnu/lxc/sys/kernel/debug', type 'none'
            lxc 20160707201003.741 DEBUG    lxc_conf - conf.c:mount_entry:1657 - remounting /sys/kernel/security on /usr/lib/x86_64-linux-gnu/lxc/sys/kernel/security to respect bind or remount options
            lxc 20160707201003.741 DEBUG    lxc_conf - conf.c:mount_entry:1672 - (at remount) flags for /sys/kernel/security was 4110, required extra flags are 14
            lxc 20160707201003.741 DEBUG    lxc_conf - conf.c:mount_entry:1707 - mounted '/sys/kernel/security' on '/usr/lib/x86_64-linux-gnu/lxc/sys/kernel/security', type 'none'
            lxc 20160707201003.741 DEBUG    lxc_conf - conf.c:mount_entry:1657 - remounting /dev/mqueue on /usr/lib/x86_64-linux-gnu/lxc/dev/mqueue to respect bind or remount options
            lxc 20160707201003.741 DEBUG    lxc_conf - conf.c:mount_entry:1672 - (at remount) flags for /dev/mqueue was 4096, required extra flags are 0
            lxc 20160707201003.741 DEBUG    lxc_conf - conf.c:mount_entry:1681 - mountflags already was 20480, skipping remount
            lxc 20160707201003.741 DEBUG    lxc_conf - conf.c:mount_entry:1707 - mounted '/dev/mqueue' on '/usr/lib/x86_64-linux-gnu/lxc/dev/mqueue', type 'none'
            lxc 20160707201003.741 DEBUG    lxc_conf - conf.c:mount_entry:1657 - remounting /var/lib/lxd/devlxd on /usr/lib/x86_64-linux-gnu/lxc/dev/lxd to respect bind or remount options
            lxc 20160707201003.741 DEBUG    lxc_conf - conf.c:mount_entry:1672 - (at remount) flags for /var/lib/lxd/devlxd was 4096, required extra flags are 0
            lxc 20160707201003.741 DEBUG    lxc_conf - conf.c:mount_entry:1681 - mountflags already was 4096, skipping remount
            lxc 20160707201003.741 DEBUG    lxc_conf - conf.c:mount_entry:1707 - mounted '/var/lib/lxd/devlxd' on '/usr/lib/x86_64-linux-gnu/lxc/dev/lxd', type 'none'
            lxc 20160707201003.741 ERROR    lxc_utils - utils.c:safe_mount:1692 - Permission denied - Failed to mount /var/lib/lxd/devices/mycontainername/unix.dev-net-tun onto /usr/lib/x86_64-linux-gnu/lxc/dev/net/tun
            lxc 20160707201003.741 ERROR    lxc_conf - conf.c:mount_entry:1650 - Permission denied - failed to mount '/var/lib/lxd/devices/mycontainername/unix.dev-net-tun' on '/usr/lib/x86_64-linux-gnu/lxc/dev/net/tun'
            lxc 20160707201003.741 ERROR    lxc_conf - conf.c:lxc_setup:3731 - failed to setup the mount entries for 'mycontainername'
            lxc 20160707201003.741 ERROR    lxc_start - start.c:do_start:833 - failed to setup the container
            lxc 20160707201003.741 ERROR    lxc_sync - sync.c:__sync_wait:57 - An error occurred in another process (expected sequence number 3)
            lxc 20160707201003.741 WARN     lxc_conf - conf.c:lxc_delete_network:2909 - failed to remove interface 18 'eth0'
            lxc 20160707201003.741 ERROR    lxc_start - start.c:__lxc_start:1353 - failed to spawn 'mycontainername'
            lxc 20160707201003.780 INFO     lxc_conf - conf.c:run_script_argv:367 - Executing script '/usr/share/lxcfs/lxc.reboot.hook' for container 'mycontainername', config section 'lxc'
            lxc 20160707201004.282 INFO     lxc_conf - conf.c:run_script_argv:367 - Executing script '/usr/bin/lxd callhook /var/lib/lxd 2 stop' for container 'mycontainername', config section 'lxc'
            lxc 20160707201004.430 WARN     lxc_commands - commands.c:lxc_cmd_rsp_recv:172 - command get_cgroup failed to receive response
            lxc 20160707201004.430 WARN     lxc_commands - commands.c:lxc_cmd_rsp_recv:172 - command get_cgroup failed to receive response
            lxc 20160707201004.434 INFO     lxc_confile - confile.c:config_idmap:1500 - read uid map: type u nsid 0 hostid 100000 range 65536
            lxc 20160707201004.434 INFO     lxc_confile - confile.c:config_idmap:1500 - read uid map: type g nsid 0 hostid 100000 range 65536
            lxc 20160707201017.724 INFO     lxc_confile - confile.c:config_idmap:1500 - read uid map: type u nsid 0 hostid 100000 range 65536
            lxc 20160707201017.724 INFO     lxc_confile - confile.c:config_idmap:1500 - read uid map: type g nsid 0 hostid 100000 range 65536
            lxc 20160707201017.734 INFO     lxc_confile - confile.c:config_idmap:1500 - read uid map: type u nsid 0 hostid 100000 range 65536
            lxc 20160707201017.734 INFO     lxc_confile - confile.c:config_idmap:1500 - read uid map: type g nsid 0 hostid 100000 range 65536

```
```
name: mycontainername
profiles:
- default
config:
  volatile.base_image: uidhere
  volatile.eth0.hwaddr: machere
  volatile.eth0.name: eth0
  volatile.last_state.idmap: '[{"Isuid":true,"Isgid":false,"Hostid":100000,"Nsid":0,"Maprange":65536},{"Isuid":false,"Isgid":true,"Hostid":100000,"Nsid":0,"Maprange":65536}]'
devices:
  root:
    path: /
    type: disk
  tun:
    path: /dev/net/tun
    type: unix-char
ephemeral: false

```

**Comment by cryptofuture** [comment]
I created new container, and tun works with new container. After container reboot too!

Also in process I found out that I had old certificate. Fixed that with removing `~/.confif/lxc` dir.
```
Get https://images.linuxcontainers.org:8443/1.0/images?recursion=1: x509: certificate signed by unknown authority
```
Container I had and still have issue, with unworkable tun (non needed now, could use tun with fresh one) was active since May, 2015 and was upgraded to xenial from trusty some time ago.
